### DISCUSSION TITLE
Cookies with JS pb_hooks

### DISCUSSION THREAD

**Discussion by pietru2004** [pocketbase/pocketbase#3447] [Q&A] [answered]
Hello I am experimenting with PB Rendering templates.

Question:
Is there way to have user recognized as authorized when they send html form without JS?
If answer for above is yes could there be small sniped in docs for setting cookies via PB extend with JS?(Yes I know that types reference shows that context has `cookie(name)`, `cookies` and `setCookie(cookie)`)

Reason:
I want to try to make website without HTML and CSS if possible, cause some people disable JS and some browser security levaels also do that...

Aditional Context:
I have seen that cookies are available... Also I don't want to make PB run any additional http requests...
I realize I am probably blind and didn't see how to do it on wiki.
I am suspecting `$app.dao().findAuthRecordByToken()`, but from other side it feels like to many args in docs for it to be this...

**Comment by ganigeorgiev** [comment]
Keep in mind that PocketBase APIs are fully stateless and there are no sessions out of the box and usually the requests require a valid `Authorization` header in order to load the auth state.

But if you want to use cookies, here is an example (_I haven't tested it_):

```js
// pb_hooks/main.pb.js

// Registers a middleware that will try to load the current request
// record auth state from a cookie.
routerUse((next) => {
    return (c) => {
        try {
            const token = c.request().cookie("pb_auth")?.value

            if (token) {
                const record = $app.dao().findAuthRecordByToken(
                    token,
                    $app.settings().recordAuthToken.secret,
                )
                c.set("authRecord", record)
            }
        } catch(_) {}

        return next(c)
    }
})

// Hook to write a cookie to the response after authentication.
// 
// Note: usually e.httpContext.setCookie(http.Cookie) should be used
// but the JSVM currently doesn't have bindings for the http.Cookie creation
// (I'll consider adding it with the next release),
// so instead we construct a raw cookie string and write it directly as response header.
onRecordAuthRequest((e) => {
    let rawCookie = `pb_auth=${e.token}; Max-Age=${$app.settings().recordAuthToken.duration}; Path=/; SameSite=Strict; Secure; HttpOnly`;

    e.httpContext.response().header().add("Set-Cookie", rawCookie);

    // ideally in the next release the above can be replaced with something like:
    // e.httpContext.setCookie(new Cookie({
    //     name:     "pb_auth",
    //     value:    e.token,
    //     secure:   true,
    //     sameSite: "Strict",
    //     httpOnly: true,
    //     maxAge:   $app.settings().recordAuthToken.duration,
    //     path:     "/",
    // }))
})
```

**Reply by pietru2004** [reply]
Ok gona check/test it when I get moment...

**Reply by pietru2004** [reply]
I have checked and it works... now I need to try to get custom route for auth that will redirect after login...

**Reply by ganigeorgiev** [reply, marked as answer]
Hm, in this case you can try something like:

```js
// "POST /login" route with "identity" and "password" fields as body
routerAdd("POST", "/login", (c) => {
    try {
        const users = $app.dao().findCollectionByNameOrId("users")

        const form = new RecordPasswordLoginForm($app, users);

        c.bind(form);
        // this will perform validation and will try to find an auth record matching the submitted credentials
        const authRecord = form.submit();

        const token = $tokens.recordAuthToken($app, authRecord)

        const rawCookie = `pb_auth=${token}; Max-Age=${$app.settings().recordAuthToken.duration}; Path=/; SameSite=Strict; Secure; HttpOnly`;

        c.response().header().add("Set-Cookie", rawCookie);
    } catch (err) {
        // maybe redirect to an error page?
        return c.redirect(400, "/error");
    }

    return c.redirect(200, "/home");
})
```

**Reply by ganigeorgiev** [reply]
Or alternatively, you can also try returning a redirect response in the above `onRecordAuthRequest` hook (aka. `e.httpContext.redirect(...)`), but note that this may break the regular JSON user auth API if some other part of your application uses it.

**Reply by pietru2004** [reply]
to make sure it won't get broken I made it like this
```js
routerAdd("post", "/login", (c) => {
    let hide_form = false;
    try {
        const users = $app.dao().findCollectionByNameOrId("users")

        const form = new RecordPasswordLoginForm($app, users);

        c.bind(form);
        // this will perform validation and will try to find an auth record matching the submitted credentials
        const authRecord = form.submit();

        const token = $tokens.recordAuthToken($app, authRecord)

        const rawCookie = `pb_auth=${token}; Max-Age=${$app.settings().recordAuthToken.duration}; Path=/; SameSite=Strict; Secure; HttpOnly`;

        c.response().header().add("Set-Cookie", rawCookie);
        hide_form=true
    } catch (err) {
        // maybe redirect to an error page?
        //return c.redirect(400, "/error");
    }
    const html = $template.loadFiles(
        `${__hooks}/views/base/layout.html`,
        `${__hooks}/views/base/header.html`,
        `${__hooks}/views/base/footer.html`,
        `${__hooks}/views/auth/login_page.html`,
    ).render({
        "tittle":"Authorize",
        "hide_form":hide_form,
    })

    return c.html(200, html)
})
```
I guess now I can just check for both post and get is user logged and if is then just only show link to go to home page

**Reply by pietru2004** [reply]
so thank you for help... and I would love to see all that stuff in docs...
I know there is `PocketBase JSVM reference`(/jsvm), but it is bit harder to navigate than docs from /docs

**Reply by vanevalverdes** [reply]
Hi! 

First, thanks for the help. This method works perfect for me. But, now, I cant logout the users in JS pocketbase as a framework. I tried to use
`$app.authStore.clear();`
But its not working. I tried to find any method in the JSVM or the documentation page, but these describe only the method for SDK.

Thank you again.

**Reply by ganigeorgiev** [reply]
@vanevalverdes I'm confused. When using JS pb_hooks there is no `authStore`, this is available only on the client-side SDKs. `$app` in the above is the running PocketBase instance.

If you use the above, to logout your users you can create a custom `/logout` endpoint that will clear/invalidate your custom cookie (eg. set cookie with expired date) when returning the response to the user.

**Reply by pietru2004** [reply]
@vanevalverdes to logout you can set cookie max age as 0 and pb auth value to "" (empty string)

**Reply by pietru2004** [reply]
@vanevalverdes 
https://github.com/pietru2004/vid_app/blob/main/pb_hooks/auth_login.pb.js#L65

**Reply by ganigeorgiev** [reply]
_Side-note: @pietru2004 it is not required, but because modern browsers can do prefetching on GET links, to avoid accidentally signing out your users you may want to change the `/signout` request type to `POST` (this will also prevent "malicious" logouts in case someone sends you the logout link to your email and you click it)_

**Reply by pietru2004** [reply]
I'll keep it in mind, but the vid app was made 4 fun. Next project is cms based on pocketbase...

**Reply by vanevalverdes** [reply]
ok ok . Perfect, thank you!
